File: typescript/src/inspection.ts
/** A local operator's data view, without runtime writes or watermark adoption. */
import { MigrationRefused } from './errors.js'
import { checkMapProject } from './generation.js'
import type { LogicalModel } from './model.js'
import type { PlacementMap } from './placement.js'
import type { Engine } from './session.js'

export interface MigrationView {
  readonly model: LogicalModel
  readonly placement: PlacementMap
  readonly projectId: string | undefined
  engineNamed(name: string): Engine
}

export class InspectionContext implements MigrationView {
  private readonly engines: Readonly<Record<string, Engine>>
  readonly projectId: string
  constructor(readonly model: LogicalModel, readonly placement: PlacementMap,
    engines: Readonly<Record<string, Engine>>, options: { projectId: string }) {
    if (placement.contract < 4) throw new MigrationRefused('operator inspection requires a generation-bearing placement map')
    checkMapProject(placement, options.projectId)
    if (model.version !== placement.modelVersion) throw new MigrationRefused('operator inspection needs the model named by its map')
    const missing = [...new Set(Object.values(placement.groups).flatMap((spot) =>
      [spot.source, ...spot.derived].map((material) => material.engine)))].filter((name) => engines[name] === undefined).sort()
    if (missing.length > 0) throw new MigrationRefused(`operator inspection is missing engines ${JSON.stringify(missing)}`)
    this.engines = Object.freeze({ ...engines })
    this.projectId = options.projectId
    Object.freeze(this)
  }
  engineNamed(name: string): Engine {
    const engine = this.engines[name]
    if (engine === undefined) throw new MigrationRefused(`operator inspection has no engine ${name}`)
    return engine
  }
}

File: python/src/sde/inspection.py
"""A local operator's data view, separate from an application session and its watermark."""

from __future__ import annotations

from collections.abc import Mapping
from dataclasses import dataclass
from types import MappingProxyType
from typing import TYPE_CHECKING, Protocol

from .errors import MigrationRefused
from .generation import check_map_project
from .model import LogicalModel
from .placement import PlacementMap

if TYPE_CHECKING:
    from .session import Engine


class MigrationView(Protocol):
    @property
    def model(self) -> LogicalModel: ...
    @property
    def placement(self) -> PlacementMap: ...
    @property
    def engines(self) -> Mapping[str, Engine]: ...
    @property
    def project_id(self) -> str | None: ...


@dataclass(frozen=True)
class InspectionContext:
    """Read a migration's physical data with local operator connections, without runtime writes.

    The before-map continues to identify the physical copies when their native epochs differ
    during maintenance. The operation must validate those actual epochs and barriers separately.
    Construction does not adopt a runtime map or advance any watermark.
    """

    model: LogicalModel
    placement: PlacementMap
    engines: Mapping[str, Engine]
    project_id: str

    def __post_init__(self) -> None:
        if self.placement.contract < 4:
            raise MigrationRefused(
                "operator inspection requires a generation-bearing placement map"
            )
        check_map_project(self.placement, self.project_id)
        if self.model.version != self.placement.model_version:
            raise MigrationRefused("operator inspection needs the model named by its map")
        missing = sorted(
            {
                material.engine
                for group in self.placement.groups.values()
                for material in group.all()
            }
            - set(self.engines)
        )
        if missing:
            raise MigrationRefused(f"operator inspection is missing engines {missing}")
        object.__setattr__(self, "engines", MappingProxyType(dict(self.engines)))
